widgets/src/input.rs

```diff
@@ -1,108 +1,63 @@
-use ribir_core::{prelude::*, ticker::FrameMsg};
+use ribir_core::prelude::*;
+
+use crate::prelude::*;
 mod caret;
 mod caret_state;
 mod glyphs_helper;
 mod handle;
-mod selected_text;
-mod text_selectable;
-use std::ops::Range;
-
-pub use caret_state::{CaretPosition, CaretState};
-pub use selected_text::SelectedHighLightStyle;
-pub use text_selectable::TextSelectable;
-
-use crate::{
-  input::{
-    caret::*,
-    handle::{TextCaretWriter, edit_handle, edit_key_handle},
-    selected_text::*,
-    text_selectable::{SelectableText, bind_point_listener, select_key_handle},
-  },
-  layout::*,
-};
-
-#[derive(ChildOfCompose)]
-pub struct Placeholder(DeclareInit<CowArc<str>>);
-
-impl Placeholder {
-  #[inline]
-  pub fn new<const M: usize>(str: impl DeclareInto<CowArc<str>, M>) -> Self {
-    Self(str.declare_into())
-  }
-}
-
-#[derive(Clone)]
-pub struct PlaceholderStyle {
-  pub text_style: TextStyle,
-  pub foreground: Brush,
-}
-
-impl CustomStyle for PlaceholderStyle {
-  fn default_style(ctx: &impl ProviderCtx) -> Self {
-    Self {
-      foreground: Palette::of(ctx).on_surface_variant().into(),
-      text_style: TypographyTheme::of(ctx).body_medium.text.clone(),
-    }
-  }
-}
-
-#[derive(Clone, PartialEq)]
-pub struct InputStyle {
-  pub size: Option<f32>,
-}
-
-impl CustomStyle for InputStyle {
-  fn default_style(_: &impl ProviderCtx) -> Self { InputStyle { size: Some(20.) } }
-}
-
-#[derive(Clone, PartialEq)]
-pub struct TextAreaStyle {
-  pub rows: Option<f32>,
-  pub cols: Option<f32>,
-}
-
-impl CustomStyle for TextAreaStyle {
-  fn default_style(_: &impl ProviderCtx) -> Self {
-    TextAreaStyle { rows: Some(2.), cols: Some(20.) }
-  }
-}
-
-pub trait EditableText: Sized {
-  fn text(&self) -> &CowArc<str>;
-
-  fn caret(&self) -> CaretState;
 
-  fn set_text_with_caret(&mut self, text: &str, caret: CaretState);
+mod text_high_light;
+mod text_select_region;
 
-  fn writer(&mut self) -> TextCaretWriter<Self> { TextCaretWriter::new(self) }
-}
+mod text_editable;
+mod text_selectable;
 
+pub use caret::TEXT_CARET;
+pub use caret_state::{CaretPosition, CaretState};
+pub use handle::{EditableText, SelectableText};
+use text_editable::edit_text;
+pub use text_high_light::TEXT_HIGH_LIGHT;
+pub use text_select_region::*;
+
+use crate::layout::Stack;
+/// The `Input` struct is a widget that represents a text input field
+/// that displays a single line of text. if you need multi line text, use
+/// `[TextArea]`
 #[derive(Declare)]
 pub struct Input {
-  #[declare(default = TypographyTheme::of(BuildCtx::get()).body_large.text.clone())]
-  pub style: TextStyle,
   #[declare(skip)]
   text: CowArc<str>,
+
   #[declare(skip)]
   caret: CaretState,
-  #[declare(default = InputStyle::of(BuildCtx::get()).size)]
+
+  /// set the max length size of the text in the view
+  /// If None, The size can be automatically calculated
+  #[declare(default)]
   size: Option<f32>,
```

Comment:
How about removing the `size` field and allowing users to choose between using `BoxClamp` or `TextClamp` or any other fixed size container to limit the input size, which may offer more flexibility to users.